platform-sdk/docs/proposals/consensus-layer/Consensus-Layer.md

```diff
@@ -0,0 +1,438 @@
+document dictate whether raw TCP, UDP, HTTP2, gRPC, or other network protocols are used. This will be left to the design
+documents for Gossip.
+
+![Gossip](gossip-system.png)
+
+Gossip is the only part of Consensus that communicates over the network with gossip peers. When Gossip is initialized,
+it is supplied a "roster". This roster contains the full set of nodes participating in gossip, along with their metadata
+such as RSA signing keys, IP addresses, and so forth. The Gossip module decides which peers to gossip with (using
+whatever algorithm it chooses).
+
+#### Events
+
+Gossip is event-oriented, meaning that it is given events to gossip, and emits events it receives through gossip. An
+implementation of Gossip could be based on a lower-level implementation based on bytes, but at the module level, it
+works in terms of events.
+
+When the Gossip module receives events through gossip, it *may* choose to perform some deduplication before sending
+them to Event Intake, but it is not required to do so.
+
+Honest nodes will **only** send valid events in *topological order* to peers. A peer may still receive events out of
+order, because different events may arrive from different peers at different times. To support topological ordering,
+events received by Gossip are not retained by Gossip, they are simply passed directly to Event Intake, which after
+verification, also orders events in topological order and recorded durably to disk, and then send back to the Gossip
+module to be sent out to its gossip peers.
+
+**The key design principle here is that a node will only send valid events through gossip**. If invalid events are
+ever received, you may know the node that sent them to you is dishonest. Validating events increases latency at each
+"hop", but allows us to identify dishonest gossip peers and discipline them accordingly.
+
+During execution, for all nodes that are online and able to keep up, events are received "live" and processed
+immediately and re-gossiped. However, if a node is offline and then comes back online, or is starting back up after
+reconnect, it may be missing events. In this case, the node will need to ask its peers for any events it is missing.
+
+For this reason, every honest node needs to buffer some events, so when its peer asks it for events, it is able to
+give them. Events can be "non-ancient", "ancient", or "expired". Although the definition for these is subject to
+configuration, we will here use some example values. Non-ancient events are those events that are either not yet
+in a consensus round, or are in one of the last 26 consensus rounds. Expired events are those that are in a round
+that is older than the last most recent 500 consensus rounds. And ancient events are those that are in a round between
+26 and 500.
+
+The Gossip module must cache all non-expired events. Note that it must be a fixed number of rounds that defines
```

Comment:
"must" may not actually be accurate, at least with the system as it is today. Currently, a node will not keep ancient but non-expired events in memory after a restart (we only read non-ancient events from the PCES), and after a reconnect there is no guarantee that anybody has the required events that are ancient-non-expired from the reconnecting node's perspective.

Reply:
Is it possible that we get into a situation where the network halts and cannot advance? If the other parent selection algorithm isn't able to create an event because it wouldn't advance consensus, and if all nodes are in the same state, then the only way out of that state is for some event to be created and distributed. Under normal circumstances, this should never happen. But if we had a partition, and the network quiesced, and both partitions were rebooted, then we'd back up but we need some events to be gossiped to get working again. Maybe the non-ancient events are sufficient? Or do we need to support the older ones too?

Reply:
Oh, I see you addressed that in the next comment.